File: hashes.js
function findIntersection(arr1, arr2) {
  let lookup = {};
  let intersection = [];
  arr1.forEach(elem => lookup[elem] = true);
  arr2.forEach(elem => {
    if (lookup[elem]) {
      intersection.push(elem);
    }
  })

  return intersection;
}

// console.log(findIntersection([1, 2, 3, 4, 5], [0, 2, 4, 6, 8]));

function findDup(arr1) {
  let seen = {};

  for (let i = 0; i < arr1.length; i++) {
    let currentVal = arr1[i];
    if (seen[currentVal]) {
      return currentVal
    } else {
      seen[currentVal] = true;
    }
  }

  return false;
}

//console.log(findDup(["a", "b", "c", "d", "c", "e", "f"]));

var a = 97;
var charArray = {};
for (var i = 0; i<26; i++)
    charArray[String.fromCharCode(a + i)] = false;
//console.log(charArray);

function findMissingAlpha(string) {
  string = string.toLowerCase();
  for (let i = 0; i < string.length; i++) {
    charArray[string[i]] = true;
  }
  for (let key in charArray) {
    if (charArray[key] === false)
    return key
  }
}

//console.log(findMissingAlpha("the quick brown box jumps over a lazy dog"))

function firstNonDUp(string) {
  let seen = {};
  for (let i = 0; i < string.length; i++) {
    seen[string[i]] = seen[string[i]] ? seen[string[i]] += 1 : 1;
  }
  console.log(seen)
  for (let i = 0; i < string.length; i++) {
    if (seen[string[i]] === 1) {
      return string[i]
    }
  }
}

// console.log(firstNonDUp("minimum"))

File: go_data_structures/hashes.go
package main

import "fmt"

func findIntersection(slice1 []int, slice2 []int) (intersection []int) {
  lookup := map[int]bool{}
  for _, elem := range slice1 {
    lookup[elem] = true
  }
  for _, elem2 := range slice2 {
    if lookup[elem2] {
      intersection = append(intersection, elem2)
    }
  }
  return intersection
}

func findDup (strings []string) (dup string) {
  seen := map[string]bool{}

  for _, elem := range strings {
    if (seen[elem]) {
      return elem
    } else {
      seen[elem] = true
    }
  }
  return dup
}

func findMissingAlpha (string1 string) (missing string) {
  alphabet := "abcdefghijklmnopqrstuvwxyz"
  seen := map[string]bool{}
  for _, char := range string1 {
    seen[string(char)] = true
  }
  for _, alpha := range alphabet {
    if !seen[string(alpha)] {
      return string(alpha)
    }
  }
  return ""
}

func firstNonDup(string1 string) (nonDup string) {
  seen := map[string]int{}
  for _, elem := range string1 {
    if seen[string(elem)] >= 1 {
      seen[string(elem)] += 1
    } else {
      seen[string(elem)] = 1
    }
  }
  fmt.Println(seen)

  for _, elem := range string1 {
    if seen[string(elem)] == 1 {
      nonDup = string(elem)
      return nonDup
    }
  }
  return nonDup
}

func main() {
  // x := []int{1, 2, 3, 4, 5}
  // y := []int{0, 2, 4, 6, 8}
  // fmt.Println(findIntersection(x, y))

  // strings := []string{"a", "b", "c", "d", "c", "e", "f"}
  // fmt.Println(findDup(strings))

  //fmt.Println(findMissingAlpha("the quick brown box jumps over a lazy dog"))

  fmt.Println(firstNonDup("minimum"))
}

